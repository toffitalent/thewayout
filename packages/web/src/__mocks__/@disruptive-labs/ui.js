const actual = jest.requireActual('@disruptive-labs/ui');

const showToast = jest.fn();

actual.Toast.show = showToast;
actual.Toast.danger = showToast;
actual.Toast.info = showToast;
actual.Toast.success = showToast;
actual.Toast.warning = showToast;

module.exports = {
  ...actual,
  showToast,
};
