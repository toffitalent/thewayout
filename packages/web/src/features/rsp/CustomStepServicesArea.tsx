import {
  Button,
  Combobox,
  ComboboxInput,
  ComboboxLabel,
  ComboboxMenu,
  Flex,
  Tag,
  TagButton,
  TagLabel,
  Text,
  useWizardContext,
  WizardStep,
} from '@disruptive-labs/ui';
import XIcon from '@disruptive-labs/ui/dist/icons/X';
import { ChangeEvent, ReactNode, useRef, useState } from 'react';
import { counties } from '@two/shared';
import styles from './CustomStepServicesArea.module.scss';
import type { WizardRspData } from './rspData';

interface ServiceAreaFormProps {
  data: string[];
  label: ReactNode;
  name: string;
  buttonLabel?: string;
  isCreator?: boolean;
  onSubmit: ({ areas }: { areas: string[] }) => void;
}

export const ServiceAreaForm = ({
  data,
  label,
  name,
  buttonLabel,
  isCreator = false,
  onSubmit,
}: ServiceAreaFormProps) => {
  const [areas, setAreas] = useState<string[]>(data);
  const [area, setArea] = useState<string>('');
  const ref = useRef<HTMLInputElement | null>(null);

  const handleRemove = (area: string) => {
    if (areas && areas.length) {
      const i = areas.indexOf(area);
      const newAreas = [...areas];
      newAreas.splice(i, 1);

      setAreas(newAreas);
    }
  };

  const isValid = () => {
    if (isCreator && !(areas || []).length) {
      return false;
    }

    if (!isCreator) {
      if (!areas.length) {
        return false;
      }
      return !(
        data.length === areas.length &&
        [...data].sort().every((el, index) => el === [...areas].sort()[index])
      );
    }

    return true;
  };

  return (
    <>
      <Combobox
        options={counties}
        onSelectedItemChange={({ selectedItem }) => {
          if (selectedItem && !(areas || []).includes(selectedItem)) {
            setAreas([...(areas || []), ...(selectedItem ? [selectedItem] : [])]);
            setArea('');
          }
          ref.current?.blur();
        }}
        selectedItem={area}
      >
        <Flex width={96} className="combobox">
          <ComboboxLabel>{label}</ComboboxLabel>
          <ComboboxInput
            name={name}
            value={area}
            width={96}
            onChange={({ target }: ChangeEvent<HTMLInputElement>) => setArea(target.value)}
            ref={ref}
          />
        </Flex>
        <ComboboxMenu className="combobox" />
      </Combobox>
      {areas && !!areas.length && (
        <Flex>
          <Text mt={10} mb={4} fontWeight="700" fontSize="xl">
            Selected Service Areas
          </Text>
          <Flex direction="row" wrap="wrap" container spacing={4}>
            {areas.map((area) => (
              <Flex item key={area}>
                <Tag size="sm" colorScheme="grey">
                  <TagLabel>{area}</TagLabel>
                  <TagButton>
                    <XIcon onClick={() => handleRemove(area)} />
                  </TagButton>
                </Tag>
              </Flex>
            ))}
          </Flex>
        </Flex>
      )}
      <Button
        colorScheme="primary"
        fluid
        width={96}
        mt={8}
        onClick={() => onSubmit({ areas })}
        disabled={!isValid()}
      >
        {buttonLabel}
      </Button>
    </>
  );
};

export const CustomStepServicesArea = (props: Partial<WizardStep>) => {
  const { next, data } = useWizardContext<WizardRspData>();

  const handleSubmit = async ({ areas }: { areas: string[] }) => {
    await next({
      ...data,
      servicesArea: areas,
    });
  };

  const { fields, nextButtonLabel } = props;
  if (!fields) {
    return null;
  }
  const [servicesArea] = fields;

  return (
    <Flex className={styles.formWrapper}>
      <ServiceAreaForm
        data={data.servicesArea}
        label={servicesArea.label}
        name={servicesArea.name}
        buttonLabel={nextButtonLabel}
        isCreator
        onSubmit={handleSubmit}
      />
    </Flex>
  );
};
