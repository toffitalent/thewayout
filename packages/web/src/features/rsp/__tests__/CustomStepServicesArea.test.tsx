import { useWizardContext } from '@disruptive-labs/ui';
import userEvent from '@testing-library/user-event';
import { act, fireEvent, render, waitFor } from '@test';
import { CustomStepServicesArea } from '../CustomStepServicesArea';

const fields = [
  {
    name: 'servicesArea',
    type: 'text',
    label: 'Search Area',
  },
];
jest.mock('@disruptive-labs/ui', () => ({
  ...jest.requireActual('@disruptive-labs/ui'),
  useWizardContext: jest.fn(),
}));

describe('CustomStepServicesArea', () => {
  beforeEach(() => {
    (useWizardContext as any).mockImplementation(() => ({
      data: {},
      setData: jest.fn(),
      goTo: jest.fn(),
    }));
  });

  test('renders correctly', async () => {
    const { container, getAllByLabelText } = render(
      <CustomStepServicesArea fields={fields as any} />,
    );

    await waitFor(async () => expect(getAllByLabelText('Search Area')[0]).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });

  test('renders list', async () => {
    const { container, getAllByLabelText } = render(
      <CustomStepServicesArea fields={fields as any} />,
    );

    await userEvent.type(getAllByLabelText('Search Area')[0], 'W');

    expect(container).toMatchSnapshot();
  });

  test('renders chosen options', async () => {
    const { container, getAllByLabelText, findByRole, findByText } = render(
      <CustomStepServicesArea fields={fields as any} />,
    );

    await userEvent.type(getAllByLabelText('Search Area')[0], 'W');
    await act(async () => fireEvent.click(await findByRole('option', { name: 'Washington' })));

    expect(
      (await findByRole('combobox', { name: 'Search Area' })).getAttribute('aria-expanded'),
    ).toBe('false');
    expect(await findByText('Selected Service Areas')).toBeInTheDocument();
    expect(await findByText('Washington')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
