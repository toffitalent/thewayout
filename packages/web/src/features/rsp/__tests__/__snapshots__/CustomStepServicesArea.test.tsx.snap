// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomStepServicesArea renders chosen options 1`] = `
<div>
  <div
    class="formWrapper"
  >
    <div
      class="control"
      role="group"
    >
      <div
        class="combobox width-96"
      >
        <label
          class="label"
          for=":r2:-input"
          id=":r2:-label"
        >
          Search Area
        </label>
        <div
          class="input"
        >
          <div
            class="root primary md md width-96"
          >
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-controls=":r2:-menu"
              aria-expanded="false"
              aria-labelledby=":r2:-label"
              autocomplete="off"
              class="input"
              id=":r2:-input"
              name="servicesArea"
              role="combobox"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="menu"
        data-popper-escaped="true"
        data-popper-placement="bottom-start"
        data-popper-reference-hidden="true"
        role="presentation"
        style="position: absolute; top: 0px; left: 0px; will-change: transform, visibility; transform: translate(0px, 5px); width: 0px;"
      >
        <ul
          aria-labelledby=":r2:-label"
          class="list combobox"
          id=":r2:-menu"
          role="listbox"
        />
      </div>
    </div>
    <div
      class=""
    >
      <p
        class="text marginTop-10 marginBottom-4 fontSize-xl fontWeight-700"
      >
        Selected Service Areas
      </p>
      <div
        class="container spacing-4 flexDirection-row flexWrap-wrap"
      >
        <div
          class="item"
        >
          <span
            class="tag grey sm outline borderColor-grey bgcolor-grey color-grey"
          >
            <span
              class="label"
            >
              Washington
            </span>
            <button
              class="button"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="icon"
                focusable="false"
                stroke-width="2"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0 0h24v24H0z"
                  stroke="none"
                />
                <path
                  d="M18 6L6 18M6 6l12 12"
                />
              </svg>
            </button>
          </span>
        </div>
      </div>
    </div>
    <button
      class="button md solid c-primary fluid width-96 marginTop-8"
      type="button"
    />
  </div>
</div>
`;

exports[`CustomStepServicesArea renders correctly 1`] = `
<div>
  <div
    class="formWrapper"
  >
    <div
      class="control"
      role="group"
    >
      <div
        class="combobox width-96"
      >
        <label
          class="label"
          for=":r0:-input"
          id=":r0:-label"
        >
          Search Area
        </label>
        <div
          class="input"
        >
          <div
            class="root primary md md width-96"
          >
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-controls=":r0:-menu"
              aria-expanded="false"
              aria-labelledby=":r0:-label"
              autocomplete="off"
              class="input"
              id=":r0:-input"
              name="servicesArea"
              role="combobox"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="menu"
        data-popper-escaped="true"
        data-popper-placement="bottom-start"
        data-popper-reference-hidden="true"
        role="presentation"
        style="position: absolute; top: 0px; left: 0px; will-change: transform, visibility; transform: translate(0px, 5px); width: 0px;"
      >
        <ul
          aria-labelledby=":r0:-label"
          class="list combobox"
          id=":r0:-menu"
          role="listbox"
        />
      </div>
    </div>
    <button
      class="button md solid c-primary disabled fluid width-96 marginTop-8"
      disabled=""
      type="button"
    />
  </div>
</div>
`;

exports[`CustomStepServicesArea renders list 1`] = `
<div>
  <div
    class="formWrapper"
  >
    <div
      class="control"
      role="group"
    >
      <div
        class="combobox width-96"
      >
        <label
          class="label"
          for=":r1:-input"
          id=":r1:-label"
        >
          Search Area
        </label>
        <div
          class="input"
        >
          <div
            class="root primary md md width-96"
          >
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-controls=":r1:-menu"
              aria-expanded="true"
              aria-labelledby=":r1:-label"
              autocomplete="off"
              class="input"
              id=":r1:-input"
              name="servicesArea"
              role="combobox"
              type="text"
              value="W"
            />
          </div>
        </div>
      </div>
      <div
        class="menu isOpen"
        data-popper-escaped="true"
        data-popper-placement="bottom-start"
        data-popper-reference-hidden="true"
        role="presentation"
        style="position: absolute; top: 0px; left: 0px; will-change: transform, visibility; transform: translate(0px, 5px); width: 0px;"
      >
        <ul
          aria-labelledby=":r1:-label"
          class="list combobox"
          id=":r1:-menu"
          role="listbox"
        >
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-0"
            role="option"
          >
            Walworth
          </li>
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-1"
            role="option"
          >
            Washburn
          </li>
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-2"
            role="option"
          >
            Washington
          </li>
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-3"
            role="option"
          >
            Waukesha
          </li>
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-4"
            role="option"
          >
            Waupaca
          </li>
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-5"
            role="option"
          >
            Waushara
          </li>
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-6"
            role="option"
          >
            Winnebago
          </li>
          <li
            aria-selected="false"
            class="item"
            id=":r1:-item-7"
            role="option"
          >
            Wood
          </li>
        </ul>
      </div>
    </div>
    <button
      class="button md solid c-primary disabled fluid width-96 marginTop-8"
      disabled=""
      type="button"
    />
  </div>
</div>
`;
